src/packagedcode/freebsd.py

```diff
@@ -0,0 +1,161 @@
+
+import codecs
+import logging
+
+from commoncode import filetype
+from commoncode import fileutils
+from commoncode import saneyaml
+from packagedcode import models
+
+"""
+Handle FreeBSD ports
+per https://www.freebsd.org/cgi/man.cgi?pkg-create(8)
+"""
+
+logger = logging.getLogger(__name__)
+# import sys
+# logging.basicConfig(level=logging.DEBUG, stream=sys.stdout)
+# logger.setLevel(logging.DEBUG)
+
+
+class FreeBSDPackage(models.Package):
+    metafiles = ('+COMPACT_MANIFEST')
+    type = models.StringType(default='freebsd')
+
+    @classmethod
+    def recognize(cls, location):
+        return parse(location)
+
+
+def is_freebsd_manifest(location):
+    return (filetype.is_file(location) 
+            and fileutils.file_name(location).lower() == '+compact_manifest')
+
+
+def parse(location):
+    """
+    Return a Package object from a +COMPACT_MANIFEST file or None.
+    """
+    if not is_freebsd_manifest(location):
+        return
+    
+    with codecs.open(location, encoding='utf-8') as loc:
+        freebsd_manifest = saneyaml.load(loc)
+
+    # a +COMPACT_MANIFEST is at the root of a FreeBSD package (port)
+    base_dir = fileutils.parent_directory(location)
+    return build_package(freebsd_manifest, base_dir)
+
+
+def build_package(package_data, base_dir=None):
+    """
+    Return a Package object from a package_data mapping (from a 
+    +COMPACT_MANIFEST file or similar) or None.
+    """
+    # construct the package
+    package = FreeBSDPackage()
+
+    # mapping of top level package.json items to the Package object field name
+    plain_fields = [
+        ('name', 'name'),
+        ('version', 'version'),
+        ('www', 'homepage_url'),
+        ('desc', 'description'),
+        ('categories', 'keywords'),
+    ]
+
+    for source, target in plain_fields:
+        value = package_data.get(source)
+        if value:
+            if isinstance(value, basestring):
+                value = value.strip()
+            if value:
+                setattr(package, target, value)
+
+    # mapping of top level +COMPACT_MANIFEST items to a function accepting as
+    # arguments the package.json element value and returning an iterable of key,
+    # values Package Object to update
+    field_mappers = [
+        ('maintainer', maintainer_mapper),
+        ('licenses', license_mapper),
+        ('origin', origin_mapper),
+        ('arch', arch_mapper),
+    ]
+
+    for source, func in field_mappers:
+        logger.debug('parse: %(source)r, %(func)r' % locals())
+        value = package_data.get(source) or None
+        if value:
+            func(value, package)
+
+    return package
+
+
+def license_mapper(licenses, package):
+    """
+    Update package licensing and return package. Licensing structure for FreeBSD
+    packages is a list of (non-scancode) license keys.
+    """
+    if isinstance(licenses, list):
+        lics = [l.strip() for l in licenses if l and l.strip()]
+        lics = ', '.join(lics)
```

Comment:
We would need to take into account the licenselogic for this too. This can be dealt with afterwards and I posted a comment
here https://github.com/nexB/scancode-toolkit/issues/1073#issuecomment-392041079